### ISSUE TITLE
afpd crash when moving file with shortened file name into subdir

### ISSUE THREAD

**Issue by rdmark** [Netatalk/Netatalk#150]
When moving a file that has been shortened from a longer file name to the max 31 chars on HFS into a subdir on the AFP share from within classic Mac OS, afpd will crash.

Test steps:
1. Create a file that is >31 chars long in the shared dir on the server running Netatalk, e.g. "touch 12345678901234567890123456789012"
2. Connect to the AFP share from a classic Mac OS host (I've tested with system 7.6.1, 7.5.5, 7.1.1, both real Macs and Basilisk II)
3. Move the above file into a subdir in the AFP share (and note how the file has been automatically renamed "12345678901234567890123456#3065")
4. Crash!
5. Once afpd starts back up again, you can access the AFP share fine, and the file you moved has been moved to the new location
6. However, every time you move this file again into a deeper dir level, the same crash will appear again

Tested with Netatalk 2.2.6

Does *not* reproduce when following the exact same steps with Netatalk 3.1.12

**Comment by rdmark** [comment]
log excerpt
```
Jan 09 20:41:29.282844 afpd[11317] {cnid_dbd.c:290} (D10:CNID): send_packet: {done}
Jan 09 20:41:29.283724 afpd[11317] {cnid_dbd.c:355} (D10:CNID): dbd_rpc: {done}
Jan 09 20:41:29.284133 afpd[11317] {cnid_dbd.c:409} (D10:CNID): transmit: {done}
Jan 09 20:41:29.284365 afpd[11317] {cnid_dbd.c:893} (D5:CNID): cnid_dbd_update: updated
Jan 09 20:41:29.284643 afpd[11317] {directory.c:561} (D5:AFPDaemon): dirlookup(did: 12706): START
Jan 09 20:41:29.284830 afpd[11317] {dircache.c:345} (D5:AFPDaemon): dircache(cnid:12706): {not in cache}
Jan 09 20:41:29.284951 afpd[11317] {directory.c:613} (D5:AFPDaemon): dirlookup(did: 12706): querying CNID database
Jan 09 20:41:29.285310 afpd[11317] {cnid_dbd.c:674} (D5:CNID): cnid_dbd_resolve: resolving CNID: 12706
Jan 09 20:41:29.285588 afpd[11317] {cnid_dbd.c:228} (D10:CNID): write_vec: wrote 40 bytes
Jan 09 20:41:29.285773 afpd[11317] {cnid_dbd.c:290} (D10:CNID): send_packet: {done}
Jan 09 20:41:29.285994 afpd[11317] {cnid_dbd.c:355} (D10:CNID): dbd_rpc: {done}
Jan 09 20:41:29.286153 afpd[11317] {cnid_dbd.c:409} (D10:CNID): transmit: {done}
Jan 09 20:41:29.286273 afpd[11317] {cnid_dbd.c:699} (D5:CNID): cnid_dbd_resolve: resolved did: 12672, name: '01234567890123456789012345#31A2'
Jan 09 20:41:29.286462 afpd[11317] {directory.c:631} (D5:AFPDaemon): dirlookup(did: 12706): recursion for did: 12672
Jan 09 20:41:29.286612 afpd[11317] {directory.c:561} (D5:AFPDaemon): dirlookup(did: 12672): START
Jan 09 20:41:29.286781 afpd[11317] {dircache.c:335} (D5:AFPDaemon): dircache(cnid:12672): {modified:"Anarchie-160"}
Jan 09 20:41:29.286934 afpd[11317] {directory.c:1023} (D5:AFPDaemon): dir_remove(did:12672,'Anarchie-160'): {removing}
Jan 09 20:41:29.287088 afpd[11317] {dircache.c:542} (D5:AFPDaemon): dircache(did:12672,"Anarchie-160"): {removed}
Jan 09 20:41:29.287239 afpd[11317] {directory.c:613} (D5:AFPDaemon): dirlookup(did: 12672): querying CNID database
Jan 09 20:41:29.287496 afpd[11317] {cnid_dbd.c:674} (D5:CNID): cnid_dbd_resolve: resolving CNID: 12672
Jan 09 20:41:29.287713 afpd[11317] {cnid_dbd.c:228} (D10:CNID): write_vec: wrote 40 bytes
Jan 09 20:41:29.287881 afpd[11317] {cnid_dbd.c:290} (D10:CNID): send_packet: {done}
Jan 09 20:41:29.288099 afpd[11317] {cnid_dbd.c:355} (D10:CNID): dbd_rpc: {done}
Jan 09 20:41:29.288271 afpd[11317] {cnid_dbd.c:409} (D10:CNID): transmit: {done}
Jan 09 20:41:29.288423 afpd[11317] {cnid_dbd.c:699} (D5:CNID): cnid_dbd_resolve: resolved did: 2, name: 'Anarchie-160'
Jan 09 20:41:29.288578 afpd[11317] {directory.c:631} (D5:AFPDaemon): dirlookup(did: 12672): recursion for did: 2
Jan 09 20:41:29.288727 afpd[11317] {directory.c:561} (D5:AFPDaemon): dirlookup(did: 2): START
Jan 09 20:41:29.288879 afpd[11317] {directory.c:706} (D5:AFPDaemon): dirlookup(did: 2): RESULT: pdid: 1, path: "/home/pi/afpshare"
Jan 09 20:41:29.289034 afpd[11317] {directory.c:647} (D5:AFPDaemon): dirlookup(did: 12672): stating "/home/pi/afpshare/Anarchie-160"
Jan 09 20:41:29.289225 afpd[11317] {dircache.c:494} (D5:AFPDaemon): dircache(did:12672,'Anarchie-160'): {added}
Jan 09 20:41:29.289382 afpd[11317] {directory.c:706} (D5:AFPDaemon): dirlookup(did: 12672): RESULT: pdid: 2, path: "/home/pi/afpshare/Anarchie-160"
Jan 09 20:41:29.289506 afpd[11317] {directory.c:647} (D5:AFPDaemon): dirlookup(did: 12706): stating "/home/pi/afpshare/Anarchie-160/01234567890123456789012345#31A2"
Jan 09 20:41:29.289707 afpd[11317] {directory.c:696} (D5:AFPDaemon): dirlookup(did: 12706) {exit_error: AFPERR_BADTYPE}
Jan 09 20:41:29.289859 afpd[11317] {cnid_dbd.c:674} (D5:CNID): cnid_dbd_resolve: resolving CNID: 12706
Jan 09 20:41:29.290067 afpd[11317] {cnid_dbd.c:228} (D10:CNID): write_vec: wrote 40 bytes
Jan 09 20:41:29.290243 afpd[11317] {cnid_dbd.c:290} (D10:CNID): send_packet: {done}
Jan 09 20:41:29.290463 afpd[11317] {cnid_dbd.c:355} (D10:CNID): dbd_rpc: {done}
Jan 09 20:41:29.290623 afpd[11317] {cnid_dbd.c:409} (D10:CNID): transmit: {done}
Jan 09 20:41:29.290742 afpd[11317] {cnid_dbd.c:699} (D5:CNID): cnid_dbd_resolve: resolved did: 12672, name: '01234567890123456789012345#31A2'
Jan 09 20:41:29.290969 afpd[11317] {fault.c:122} (S:Default): ===============================================================
Jan 09 20:41:29.291125 afpd[11317] {fault.c:123} (S:Default): INTERNAL ERROR: Signal 11 in pid 11317 (2.2.6)
Jan 09 20:41:29.291240 afpd[11317] {fault.c:124} (S:Default): ===============================================================
Jan 09 20:41:29.292329 afpd[11317] {fault.c:96} (S:Default): BACKTRACE: 0 stack frames:
Jan 09 20:41:29.294772 afpd[10633] {main.c:493} (D5:AFPDaemon): main: IPC request from child[11317]
Jan 09 20:41:29.295006 afpd[10633] {main.c:423} (D10:AFPDaemon): main: polling 1 fds
```

**Comment by rdmark** [comment]
Thanks to the efforts of @christopherkobayashi we tracked down the Netatalk 3 patch, which was merged back in 2011 but never backported to 2.2: https://github.com/Netatalk/Netatalk/commit/05bb4cfbd3dc7e3061151c963650a126c9e412b8

**Comment by rdmark** [comment]
The patch has been merged to branch-netatalk-2-2
